src/controllers/controllerengine.cpp

```diff
@@ -1096,6 +1095,23 @@ bool ControllerEngine::evaluateWithReturn(const QFileInfo& scriptFile, QScriptVa
     scriptCode.append('\n');
     input.close();
 
+    return evaluateWithReturn(scriptCode, filename, outValue);
+}
+
+/* -------- ------------------------------------------------------
+   Purpose: Evaluate javascript
+   Input:   Script code, can contain newlines.  Filename can be empty if not
+            applicable
+   Output:  false if the script file has errors or doesn't exist, and the result
+            of the script if outValue is not nullptr.
+   -------- ------------------------------------------------------ */
+bool ControllerEngine::evaluateWithReturn(const QString& scriptCode,
```

Comment:
These changes to ControllerEngine are likely going to have to be redone from scratch for master.

Reply:
I'm not comfortable rushing this new feature into a beta branch, especially considering that it depends on already outdated code. I don't like that a C++ file has to be written for every specific controller mapping with this approach. If it helps you write this mapping, that's fine, but I think we should remove all C++ changes before merging this. I don't really want to work on this new feature while we have a major refactoring of the same part of the code in progress. If you want to discuss how to fit testing into the new controller system, let's discuss that on Zulip before rushing a feature into a beta branch.

Reply:
they're minor, so ok

Reply:
it looks like evaluateCodeString is already the thing I want -- we could commit just the js to 2.3 and commit the js plus tests to master.  or, I could target this to master and then backport just the js to 2.3

Reply:
I'm not convinced this mixing of C++ and JS is a good approach to testing scripts. I also don't think we should be investing any more effort in the legacy mapping system right now; let's work to build a better system instead. If we add a framework for testing controller scripts, IMO it should be done with self-contained JS files. We can't really do that well right with the legacy mapping system having all mapping files in one big, messy folder. Let's look at common practices for JS testing frameworks and carefully consider the design before building our own.

Reply:
Let's discuss this [on Zulip](https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping/topic/testing.20controller.20scripts).

Reply:
When a new controller system exists, I'll use it.  For now, the only way I was able to rewrite the effects system was to write these tests.  The amount of change to the legacy infrastructure is quite minimal.

Reply:
For now, what do you think of my proposal to port this to master, and then backport just the controller script to 2.3?